Add clock_configure_mhz function

will-v-pi opened the pull request:
This adds a `clock_configure_mhz` function, which trades off accuracy compared to `clock_configure` (only gets to +/- 1MHz) for reduced code size as it doesn't need 64 bit division.

It takes the source and target frequencies in MHz as 16-bit numbers, to ensure they fit within the 32-bit division.

This is useful for example if you want to calibrate the ROSC using an XOSC and then run off the ROSC, in which case you don't need a very accurate frequency as the ROSC frequency isn't accurate.

will-v-pi committed:
Add clock_configure_mhz

Trades off accuracy (only gets to +/- 1MHz) for reduced code size as it doesn't need 64 bit division

kilograham commented:
As per new code in #2225 we should catch cases where the divider between 1 and 2.

kilograham commented on src/rp2_common/hardware_clocks/include/hardware/clocks.h:
```diff
@@ -282,6 +282,22 @@ typedef clock_num_t clock_handle_t;
+ *  \ingroup hardware_clocks
+ *
+ * This function differs from clock_configure in that it does not configure the clocks as accurately,
+ * but therefore doesn't need to bring in 64-bit division functions, reducing the code size.
```
reducing code size if 64-bit division is not otherwise used by the application

lurch commented:
> It takes the source and target frequencies in MHz as 16-bit numbers

I guess that means you can't overclock the Pico to 66 GHz? :rofl:

lurch commented:
See also #2255

will-v-pi committed:
Review fixup

Fixup for dividers 1.0-2.0 on RP2040

Improve docs

will-v-pi commented:
> > It takes the source and target frequencies in MHz as 16-bit numbers
> 
> I guess that means you can't overclock the Pico to 66 GHz? 🤣

That's better than the normal clock_configure, which only lets you overclock to 4.3GHz 🤣

will-v-pi commented:
@kilograham do we want this function, or should I close this PR?

kilograham commented:
no, it is good; i didn't look in detail at the accuracy; does it at least get you near to the right answer, vs 1Mhz away ;-)

will-v-pi commented:
> no, it is good; i didn't look in detail at the accuracy; does it at least get you near to the right answer, vs 1Mhz away ;-)

If the divider is an integer (ie `(src_freq << CLOCKS_CLK_GPOUT0_DIV_INT_LSB) / freq` is an integer) then it'll always get the exact frequency, and from having a play I can't find any frequency combinations where that isn't an integer, so I think it'll generally always be correct? If there is a case where that divider is a non-integer then it's rounded down, so you'll get a slightly higher frequency than requested

kilograham commented on src/rp2_common/hardware_clocks/include/hardware/clocks.h:
```diff
@@ -318,6 +318,31 @@ typedef clock_num_t clock_handle_t;
  */
 bool clock_configure(clock_handle_t clock, uint32_t src, uint32_t auxsrc, uint32_t src_freq, uint32_t freq);
 
+/*! \brief Configure the specified clock to +/- 1MHz
```
Actually could we clarify what this means per our discussion

will-v-pi replied:
I think this is correct

```suggestion
/*! \brief Configure the specified clock with 1MHz accuracy
```

will-v-pi committed:
Update src/rp2_common/hardware_clocks/include/hardware/clocks.h

lurch commented on src/rp2_common/hardware_clocks/include/hardware/clocks.h:
```diff
@@ -318,6 +318,31 @@ typedef clock_num_t clock_handle_t;
+ * \param src The main clock source, can be 0.
+ * \param auxsrc The auxiliary clock source, which depends on which clock is being set. Can be 0
+ * \param src_freq_mhz Frequency of the input clock source in MHz
+ * \param freq_mhz Requested frequency in MHz
```
Looks like the doxygen is missing a `\return` ?

will-v-pi replied:
```suggestion
 * \param freq_mhz Requested frequency in MHz
 * \return true if the clock is updated, false if freq > src_freq
```

lurch commented on src/rp2_common/hardware_clocks/clocks.c:
```diff
@@ -123,6 +123,26 @@ bool clock_configure(clock_handle_t clock, uint32_t src, uint32_t auxsrc, uint32
+    uint32_t actual_freq = (uint32_t) ((((uint32_t) src_freq_mhz) << CLOCKS_CLK_GPOUT0_DIV_INT_LSB) / div) * MHZ;
+
+    clock_configure_internal(clock, src, auxsrc, actual_freq, div);
+    // Store the configured frequency
```
What does this comment mean? (and is it referring to the line below or the line above the comment?)

will-v-pi replied:
I’m not entirely sure (it was copied from the function above) - it’s related to the line above it as that’s what stores the configured frequency

lurch replied:
Perhaps it'd be less confusing if the comment was moved _above_ the line it's referring to, as is done conventionally? :slightly_smiling_face:

will-v-pi replied:
I've put it in the correct place now, which is inside the `clock_configure_internal` function

will-v-pi committed:
Add return to doxygen

kilograham commented on src/rp2_common/hardware_clocks/include/hardware/clocks.h:
```diff
@@ -318,6 +318,32 @@ typedef clock_num_t clock_handle_t;
+ * \param freq_mhz Requested frequency in MHz
+ * \return true if the clock is updated, false if freq > src_freq
+ */
+bool clock_configure_mhz(clock_handle_t clock, uint32_t src, uint32_t auxsrc, uint16_t src_freq_mhz, uint16_t freq_mhz);
```
hard for me to eyeball whether clock_configure_mhz(N) always configures the same clock as clock_config(N*1000000)... (and returns the same value) ... based on that we may want to rename to clock_configure_approx_mhz() and/or update comments (the current comment implies it doesn't, but with the current name i;d expect it to - we probide _hz, _khz_ _mhz functions elsewhere which do exactly the same thing)

will-v-pi replied:
I can confirm for N in range 1-150, they both configure to the same frequency

lurch replied:
Is it also true for frequencies up to 200MHz, which is the "official" max frequency for RP2040?

will-v-pi replied:
Yes, true for those too

will-v-pi committed:
Merge branch 'develop' into clock-configure-mhz

kilograham commented on src/rp2_common/hardware_clocks/clocks.c:
```diff
@@ -123,6 +123,26 @@ bool clock_configure(clock_handle_t clock, uint32_t src, uint32_t auxsrc, uint32
     return true;
 }
 
+bool clock_configure_mhz(clock_handle_t clock, uint32_t src, uint32_t auxsrc, uint16_t src_freq_mhz, uint16_t freq_mhz) {
```
actually i think i'd rather these be uint32_t - 0-65535 is not a magic range, and at first glance we don't rely on these being 16 bit anywhere.

If we want a sensible maximum we'd enforce that with an IF

(note we generally don't use smaller fields unless it explicitly matches some hardware field)

will-v-pi replied:
It has to fit in 16-bits for this line, as it gets shifted by 16:
```
uint32_t div = (uint32_t)((((uint32_t) src_freq_mhz) << CLOCKS_CLK_GPOUT0_DIV_INT_LSB) / freq_mhz);
```

lurch replied:
`CLOCKS_CLK_GPOUT0_DIV_INT_LSB` is 16 for RP2350, but only 8 for RP2040.
(clock dividers are 24.8 on RP2040 but 16.16 on RP2350)

will-v-pi replied:
Ok, I've changed it to take a `uint32_t`, and added an assert and return if it doesn't fit in `CLOCKS_CLK_GPOUT0_DIV_INT_BITS`

will-v-pi committed:
Take uint32_t, and add explicit check that it fits in CLOCKS_CLK_GPOUT0_DIV_INT_BITS

This is because it could technically take up to 24-bits on RP2040 as the DIV_INT_BITS is larger, although it's unlikely anyone would need >65535MHz

will-v-pi committed:
Tidy up `// Store the configured frequency` comments

The storing actually happens in `clock_configure_internal`

lurch commented on src/rp2_common/hardware_clocks/clocks.c:
```diff
@@ -123,12 +123,17 @@ bool clock_configure(clock_handle_t clock, uint32_t src, uint32_t auxsrc, uint32
     if (freq_mhz > src_freq_mhz)
         return false;
 
+    assert(src_freq_mhz <= (uint32_t)(CLOCKS_CLK_GPOUT0_DIV_INT_BITS >> CLOCKS_CLK_GPOUT0_DIV_INT_LSB));
```
Should this be using `invalid_params_if` rather than a manual `assert` ?

will-v-pi replied:
`clock_configure` which this was based on also uses `assert` and a return to validate parameters, so I'll stick with that in case there is a reason for it - if we want to change that behaviour that could come in a separate PR (eg to `invalid_params_if_and_return`)

lurch replied:
Okay :+1:
